postprocessing/science/rise_time_from_xdmf.py

```diff
@@ -22,44 +31,107 @@ def counting_loop(row, array, threshold=0.01):
         if array[row,j] >= threshold: count += 1
     return count
 
-def get_rise_time(ASR_array, threshold=0.01, dt=0.1, events=1):
+@jit(nopython=True)
+def get_rise_time(ASR_array, threshold=0.1, dt=0.1, events=1):
+    
+    RT = np.zeros_like(ASR_array[0:2,:])
+    mid = int(ASR_array.shape[0]/2)
+    ASR_array=np.transpose(ASR_array) # to speed up
+    
     if events == 1:
-        RT = np.zeros(ASR_array.shape[1])
-        ASR_array=np.transpose(ASR_array) # to speed up
         for i in range(ASR_array.shape[0]):
             count = counting_loop(i, ASR_array, threshold=threshold) 
-            RT[i] = (count * dt)
+            RT[0,i] = (count * dt)
+        return RT
             
     if events == 2:
-        RT = np.zeros_like(ASR_array[0:2,:])
-        mid = int(ASR_array.shape[0]/2)
-        ASR_array=np.transpose(ASR_array)
         for i in range(ASR_array.shape[0]):
             count = counting_loop(i, ASR_array[:,0:mid], threshold=threshold)      
             RT[0,i] = (count * dt) 
             count = counting_loop(i, ASR_array[:,mid:], threshold=threshold)    
             RT[1,i] = (count * dt)          
-    return RT
+        return RT
 
 def get_absolute_slip_rate(sx_object):
     SRs = sx_object.ReadData('SRs')
     SRd = sx_object.ReadData('SRd')
     ASR = np.sqrt(SRs**2 + SRd**2)
     return ASR
 
+@jit(nopython=True)
+def MedianSmoothing(input_tuple):
+    # input_tuple: (variable, xyz, indices, radius)
+    variable, xyz, indices, radius = input_tuple
+    variableSmooth = np.zeros(indices.shape[0])
+    for i, j in enumerate(indices):
+        xyz0 = xyz[j,:]
+        distances = np.sqrt(np.sum((xyz - xyz0)**2, axis=1))
+        if complicated_fault:
+            tmp_ind1 = (distances <= radius)
+            tmp_ind2 = (variable != 0)
+            tmp_ind3 = tmp_ind1 & tmp_ind2
+            non_zero_portion = np.sum(tmp_ind3)/np.sum(tmp_ind1)
+            if non_zero_portion < 0.75 and variable[j] == 0:       # improve filter for branching faults
+                variableSmooth[i] = 0
+            elif non_zero_portion < 0.35:  # remove isolated slipping elements
+                variableSmooth[i] = 0
+            else:
+                variableSmooth[i] = np.median(variable[tmp_ind3])
+        else:
+            variableSmooth[i] = np.median(variable[distances <= radius])
```

Comment:
seems like variable[distances <= radius && variable != 0] would also make sense here?

Reply:
This would lead to nans for regions with only zeros